db/commitlog/commitlog.cc

```diff
@@ -432,6 +432,8 @@ class db::commitlog::segment_manager : public ::enable_shared_from_this<segment_
     size_t max_request_controller_units() const;
     segment_id_type _ids = 0;
     std::vector<sseg_ptr> _segments;
```

Comment:
General Q: what will prevent us from issuing flushes to the entire commitlog if flush doesn't respond in time?

I agree with the patch, just worried that lack of feedback would make it cause worse behavior. IMO the lack of feedback is the main cause of trouble.

Reply:
flush_segments(0) will still generate flush callbacks for _all_ ids in all non-active segments. 

The biggest drawback I can think of is that if we've flush requested X bytes previously, causing memtable A to begin flushing, but it cannot finish because of IO queue or whatnot, if we now also need Y more bytes, we will request more flushing from maybe memtable B, even though waiting for A might have been enough. Basically, it will cause us to be more greedy (not restart from bottom). I.e. in a situation with no segments available, it should cause us to be more aggressive. But in return, we don't have to flush(0) in `allocate_segment`. Or so we hope at least.

Reply:
We had the same problem with memtable->sstable flushing (compared to the current commitlog->memtable flushing). We solved it by having the memtable->sstable flush provide feedback to the code that admits new writes. The idea is that we admit writes at the rate that the memtable flush is progressing. I'll try to write up an idea.

Reply:
That's the hard limit. The soft limit is that we only admit a byte of new memtable write if a memtable byte was written to disk. When the currently-filling memtable becomes full, the flush is complete, and we don't hit the hard limit.

Reply:
And the other part of the deal is to adjust the memtable flush rate (via the scheduling group shares) by seeing how close we are getting to the hard limit. This increases flush rate and slows down incoming writes, until the flush rate matches the incoming rate. See [blog post](https://www.scylladb.com/2018/06/12/scylla-leverages-control-theory/) by our dear @glommer, it's more about compactions, but the principle is the same.